File: iterai-js/src/diff.ts
import { createPatch } from "diff";

const logger = {
  info: (msg: string) => console.info(`[INFO] diff: ${msg}`),
};

/**
 * Generate a unified diff between two strings
 */
export function genericDiff(a: string, b: string): string {
  return createPatch("comparison", a, b, "A", "B");
}

/**
 * Print a diff with optional color output (for terminal)
 */
export function gitDiff(a: string, b: string, color: boolean = true): void {
  let diff = genericDiff(a, b);

  if (color && typeof window === "undefined") {
    // Node.js terminal colors
    const lines = diff.split("\n").map((line) => {
      if (line.startsWith("+") && !line.startsWith("+++")) {
        return `\x1b[32m${line}\x1b[0m`;
      } else if (line.startsWith("-") && !line.startsWith("---")) {
        return `\x1b[31m${line}\x1b[0m`;
      }
      return line;
    });
    diff = lines.join("\n");
  }

  logger.info(diff);
}

/**
 * Compare two plan texts
 */
export function comparePlan(planA: string, planB: string): string {
  return genericDiff(planA, planB);
}

File: iterai-py/iterai/diff.py
import difflib
import logging

logger = logging.getLogger(__name__)


def generic_diff(a: str, b: str) -> str:
    return "".join(
        difflib.unified_diff(
            a.splitlines(keepends=True),
            b.splitlines(keepends=True),
            fromfile="A",
            tofile="B",
        )
    )


def git_diff(a: str, b: str, color=True):
    diff = generic_diff(a, b)
    if color:
        lines = []
        for line in diff.split("\n"):
            if line.startswith("+") and not line.startswith("+++"):
                lines.append(f"\033[32m{line}\033[0m")
            elif line.startswith("-") and not line.startswith("---"):
                lines.append(f"\033[31m{line}\033[0m")
            else:
                lines.append(line)
        diff = "\n".join(lines)
    logger.info(diff)


def compare_plan(plan_a: str, plan_b: str) -> str:
    return generic_diff(plan_a, plan_b)
